### ISSUE TITLE
What should the API look like?

### ISSUE THREAD

**Issue by stevana** [advancedtelematic/quickcheck-state-machine#20]
Currently the user needs to provide:

- a GADT of well-typed commands
- a model
- pre-/post-conditions and a model transition function
- a generator of untyped commands without references
- a shrinker for an untyped command
- semantics of well-typed commands
- a bunch of common instances for the commands; show, read, eq, ord, functor, etc
- a function which identifies which commands return a reference/variable

Given the above, the library gives a sequential and a parallel property back to the user.

While this seems to be quite a minimal set of things we can expect from the user, there are still a couple of problems:

- the `Functor` and `Read` instances for well-typed commands can't be derived, and `Functor` requires an unsafeCoerce, see #16. These problems didn't exist before, when commands were untyped, see #9, is it worth the trouble?
- the types of the property helper functions (`sequentialProperty` and `parallelProperty`) could be simpler, perhaps using classes or records to pack up functionality?
 
Ideas on how we can fix these problems? Are there other problems? Are there better design choices?

**Comment by Danten** [comment]
Just to add my current thinking how the api for the user should look like (in the case of multiple references). So the user provides the following three types:
```
Ix :: *
Cmd :: * -> Maybe Ix -> (ix -> *) -> *
Sem :: Ix -> *
```
Here `Ix` is used to distinguish between different reference types, so in the case there is only one reference type `Ix ≅ ()`. The arguments for `Cmd resp bind ref` is as follow:
- `resp :: *` is the type of the response
- `bind :: Maybe Ix` is `Just ix` if the `Cmd` returns a reference of `ix :: Ix` type, and it is `Nothing` if no reference is returned
- `ref :: Ix -> *` is the type references.
The `Sem` type is used to store the "real" references of the system, I'm not happy that this is currently needed, but will come back to it later.

The current running example we have would look like this (with `Ix = ()`):
```
data Cmd :: * -> Maybe () -> (() -> *) -> * where
  New :: Cmd () (Just '()) ref
  Read :: ref '() -> Cmd Int Nothing ref
  Write :: ref '() -> Int -> Cmd () Nothing ref
  Inc :: ref '() -> Cmd () Nothing ref
```
The evaluation would look as follows:
```
semStep :: Cmd resp m Sem -> IO (FromMaybe () Sem m, resp)
```
Where
```
type family FromMaybe (z :: *)(s :: ix -> *)(m : Maybe ix) where
  FromMaybe z s Nothing = z
  FromMaybe z s (Just x) = s x
```
The `StateMachineModel` type gets updated to the following:
```
data StateMachineModel ix model cmd = StateMachineModel
  { precondition :: forall ref m resp. Ord1 ref => model ref -> cmd resp m ref -> Bool
  , postcondition :: forall ref m resp. Ord1 ref => model ref -> cmd resp m ref -> FromMaybe () ref m -> resp -> Property
  , transistion :: forall ref m resp. Ord1 ref => model ref -> cmd resp m ref -> FromMaybe () ref m -> resp -> model ref
  , initialModel :: forall ref. model ref
  }
```

Current benefits that I see with this is:
- `Cmd` will be a proper `Functor` (well `Functor1`) with no `unsafeCoerce`
- We don't treat any of the `resp` types as special, and don't do special casing and unsafecoerce the semantics, (so `postcondition` and `transitions` always get the `resp` but may also get the `ref` if the command binds one.

I don't know of the drawbacks yet, since we haven't experimented with this much yet.

**Comment by stevana** [comment]
A variant of the above was implemented in #33.
